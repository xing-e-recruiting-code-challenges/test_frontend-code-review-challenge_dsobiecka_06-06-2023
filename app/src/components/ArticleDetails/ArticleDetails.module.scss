@use 'styles/spacings';

.articleDetails {
  padding: spacings.$spacing-m;
  overflow: auto;
  box-sizing: border-box;
  overflow-y: auto;
  font-size: 2.4rem;
}

.title {
  font-size: 3.2rem;
  font-weight: bold;
  margin-bottom: spacings.$spacing-m;
}

.descriptionWrapper {
  display: flex;
}

.description {
  font-size: 2.4rem;
  margin-left: spacings.$spacing-m;
}

.image {
  max-width: 400px;
  width: 40%;
  height: auto;
  object-fit: contain;
}

.content {
  font-size: 1.6rem;
  margin-top: spacings.$spacing-m;
  margin-bottom: spacings.$spacing-s;
}

.url {
  font-size: 1.2rem;
}

.footer {
  margin-top: spacings.$spacing-m;
  font-size: 1.2rem;
}

.separator {
  margin: 0 spacings.$spacing-s;
}
