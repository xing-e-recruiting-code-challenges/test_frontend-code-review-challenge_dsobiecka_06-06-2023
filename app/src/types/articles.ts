type Source = {
  id: string | null;
  name: string;
};

export type Article = {
  source: any;
  author: string | null;
  title: string;
  description: string;
  url: any;
  urlToImage: any;
  publishedAt: string;
  content: string;
  id: string;
};

export type Articles = Article[];

export type ArticlesResponse = {
  articles: Article[];
};
