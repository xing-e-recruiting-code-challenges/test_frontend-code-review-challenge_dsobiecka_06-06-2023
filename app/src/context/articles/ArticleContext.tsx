import React, {createContext, useContext, useEffect, useState} from 'react';
import useFetchArticles from 'hooks/useFetchArticles';
import {Articles, Article} from 'types/articles';
import {ArticlesContextInterface} from './typings';

type ArticlesContextProviderProps = {
  children: React.ReactNode;
};

const ARTICLES_PER_PAGE = 10;

const ArticlesContext = createContext<ArticlesContextInterface | null>(null);

const ArticlesContextProvider = ({children}: ArticlesContextProviderProps) => {
  // selected article state
  const [selectedArticle, setSelectedArticle] = useState<Article | undefined>();

  // search state
  const [searchQuery, setSearchQuery] = useState<string>();

  // pagination
  const [currentPage, setCurrentPage] = useState<number>(1);
  const [paginatedArticles, setPaginatedArticles] = useState<Articles>([]);

  // request data
  const {data, isError, isLoading} = useFetchArticles(searchQuery);

  // read articles
  const [readArticlesIds, setReadArticlesIds] = useState<string[]>([]);

  const wasArticleRead = (id: string): boolean => {
    return false;
  };

  const addReadArticleById = (id: string) => {
    return;
  };

  useEffect(() => {
    if (data?.articles) {
      const firstArticleOfPage = (currentPage - 1) * ARTICLES_PER_PAGE;
      const lastArticleOfPage = currentPage * ARTICLES_PER_PAGE;
      const articlesToShow = data?.articles.slice(
        firstArticleOfPage,
        lastArticleOfPage
      );

      setPaginatedArticles(articlesToShow);
    }
  }, [data, currentPage]);

  return (
    <ArticlesContext.Provider
      value={{
        selectedArticle,
        setSelectedArticle,
        articles: paginatedArticles,
        isError,
        isLoading,
        setSearchQuery,
        addReadArticleById,
        wasArticleRead,
        pagination: {
          currentPage,
          setCurrentPage,
          totalArticles: data?.articles?.length || 0,
          articlesPerPage: ARTICLES_PER_PAGE,
        },
      }}
    >
      {children}
    </ArticlesContext.Provider>
  );
};

function useArticlesContext() {
  return useContext(ArticlesContext);
}

export default ArticlesContextProvider;
export {useArticlesContext};
