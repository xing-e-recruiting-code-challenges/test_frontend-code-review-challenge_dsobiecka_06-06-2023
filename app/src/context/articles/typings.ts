import {Article, Articles} from 'types/articles';

export interface ArticlesContextInterface {
  selectedArticle: Article | undefined;
  setSelectedArticle: Function;
  articles: Articles;
  isError: boolean;
  isLoading: boolean;
  setSearchQuery: Function;
  pagination: PaginationInterface;
  addReadArticleById: Function;
  wasArticleRead: Function;
}

export interface PaginationInterface {
  currentPage: number;
  setCurrentPage: Function;
  totalArticles: number;
  articlesPerPage: number;
}
