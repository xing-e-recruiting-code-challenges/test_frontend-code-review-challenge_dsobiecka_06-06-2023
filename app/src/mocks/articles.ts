import {ArticlesResponse} from 'types/articles';

const mockArticles: ArticlesResponse = {
  articles: [
    {
      source: {
        id: null,
        name: 'Youtube.com',
      },
      author: null,
      title:
        'Watch live: NASA to conduct historic first all-female spacewalk - Washington Post',
      description:
        'NASA astronauts Christina Koch and Jessica Meir walk outside the International Space Station to replace a damaged battery unit just months after NASA had to ...',
      url: 'https://www.youtube.com/watch?v=KGRl_-wWMD0',
      urlToImage: 'https://i.ytimg.com/vi/KGRl_-wWMD0/hqdefault_live.jpg',
      publishedAt: '2019-10-18T11:11:55Z',
      content:
        'NASA astronauts Christina Koch and Jessica Meir walk outside the International Space Station to replace a damaged battery unit just months after NASA had to cancel an all-female spacewalk because of a spacesuit-sizing problem. At a media briefing before the s… [+494 chars]',
      id: '057bca12-982e-42cb-9854-0d3ff96523e7',
    },
    {
      source: {
        id: 'the-verge',
        name: 'The Verge',
      },
      author: 'Jon Porter',
      title:
        'Samsung says fingerprint security fix is coming as early as next week - The Verge',
      description:
        'Samsung says that it plans to release a patch for its in-display fingerprint sensors as early as next week. It follows reports that the biometric security method could be compromised by certain screen protectors.',
      url: 'https://www.theverge.com/2019/10/18/20920522/samsung-galaxy-s10-note-10-fingerprint-recognition-patch-next-week-bug-biometric-security',
      urlToImage:
        'https://cdn.vox-cdn.com/thumbor/0S8WzgMFefP2HzwqwrSJi35l_0g=/0x146:2040x1214/fit-in/1200x630/cdn.vox-cdn.com/uploads/chorus_asset/file/14843648/akrales_190228_3255_0065.jpg',
      publishedAt: '2019-10-18T10:45:43Z',
      content:
        'And advises removing silicone screen protecting cases for now\r\nPhoto by Amelia Holowaty Krales / The Verge\r\nSamsung says that it plans to release a patch for its phones in-display fingerprint sensors as early as next week, after reports emerged that the biome… [+963 chars]',
      id: 'a82ae44b-2597-402e-af6a-d1e5640764b3',
    },
  ],
};

export default mockArticles;
