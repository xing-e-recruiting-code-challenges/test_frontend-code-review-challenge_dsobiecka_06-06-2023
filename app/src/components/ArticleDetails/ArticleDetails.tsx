import classNames from 'classnames';
import React from 'react';
import {useArticlesContext} from 'context/articles/ArticleContext';
import Card from 'components/ui/Card';

import styles from './ArticleDetails.module.scss';
import formatDate from 'utils/formatDate';
import {ArticlesContextInterface} from 'context/articles/typings';

type ArticleDetailsProps = {
  className?: string;
};

function ArticleDetails({className}: ArticleDetailsProps) {
  const articleDetailsClasses = classNames(className, styles.articleDetails);
  const titleClasses = classNames(styles.title, styles.h1);

  const {selectedArticle, isLoading} =
    useArticlesContext() as ArticlesContextInterface;

  if (!selectedArticle) {
    const placeholderText = !isLoading && 'Please select an article';

    return (
      <div className={articleDetailsClasses}>
        <Card>{placeholderText}</Card>
      </div>
    );
  }

  const {
    urlToImage: image,
    title,
    author,
    description,
    content,
    publishedAt,
    url,
  } = selectedArticle;

  return (
    <div className={articleDetailsClasses}>
      <Card>
        <article>
          <div className={titleClasses}>{title}</div>
          <div className={styles.descriptionWrapper}>
            {image && (
              <img
                src={image}
                loading="lazy"
                className={styles.image}
              ></img>
            )}
            <div className={styles.description}>{description}</div>
          </div>
          <div className={styles.content}>{content}</div>
          <a href={url} rel="noopener noreferrer" className={styles.url}>
            Read More
          </a>

          <footer className={styles.footer}>
            {author && (
              <>
                <span>{author}</span>
                <span className={styles.separator}>&#10625;</span>
              </>
            )}
            <span>{formatDate(publishedAt)}</span>
          </footer>
        </article>
      </Card>
    </div>
  );
}

export default ArticleDetails;
