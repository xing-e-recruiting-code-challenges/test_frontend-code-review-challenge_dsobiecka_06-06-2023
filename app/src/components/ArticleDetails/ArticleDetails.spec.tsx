import React from 'react';
import {render} from '@testing-library/react';
import {useArticlesContext} from 'context/articles/ArticleContext';
import mockArticles from 'mocks/articles';

import ArticleDetails from './ArticleDetails';

jest.mock('context/articles/ArticleContext');

const useArticlesContextMock = useArticlesContext as jest.MockedFunction<
  typeof useArticlesContext
>;
const mockArticle = mockArticles.articles[1];

describe('ArticleDetails', () => {
  beforeEach(() => {
    useArticlesContextMock.mockImplementation(() => ({
      selectedArticle: mockArticle,
      setSelectedArticle: () => {},
      articles: [],
      isError: false,
      isLoading: false,
      setSearchQuery: () => {},
      wasArticleRead: () => {},
      addReadArticleById: () => {},
      pagination: {
        currentPage: 1,
        setCurrentPage: () => {},
        totalArticles: 1,
        articlesPerPage: 10,
      },
    }));
  });

  it('should render correctly', () => {
    const {container} = render(<ArticleDetails />);

    expect(container).toMatchSnapshot();
  });
});
