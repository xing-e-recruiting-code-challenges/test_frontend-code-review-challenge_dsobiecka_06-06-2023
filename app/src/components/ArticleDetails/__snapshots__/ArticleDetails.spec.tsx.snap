// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleDetails should render correctly 1`] = `
<div>
  <div
    class="articleDetails"
  >
    <div
      class="card"
    >
      <article>
        <div
          class="title h1"
        >
          Samsung says fingerprint security fix is coming as early as next week - The Verge
        </div>
        <div
          class="descriptionWrapper"
        >
          <img
            class="image"
            loading="lazy"
            src="https://cdn.vox-cdn.com/thumbor/0S8WzgMFefP2HzwqwrSJi35l_0g=/0x146:2040x1214/fit-in/1200x630/cdn.vox-cdn.com/uploads/chorus_asset/file/14843648/akrales_190228_3255_0065.jpg"
          />
          <div
            class="description"
          >
            Samsung says that it plans to release a patch for its in-display fingerprint sensors as early as next week. It follows reports that the biometric security method could be compromised by certain screen protectors.
          </div>
        </div>
        <div
          class="content"
        >
          And advises removing silicone screen protecting cases for now
Photo by Amelia Holowaty Krales / The Verge
Samsung says that it plans to release a patch for its phones in-display fingerprint sensors as early as next week, after reports emerged that the biome… [+963 chars]
        </div>
        <a
          class="url"
          href="https://www.theverge.com/2019/10/18/20920522/samsung-galaxy-s10-note-10-fingerprint-recognition-patch-next-week-bug-biometric-security"
          rel="noopener noreferrer"
        >
          Read More
        </a>
        <footer
          class="footer"
        >
          <span>
            Jon Porter
          </span>
          <span
            class="separator"
          >
            ⦁
          </span>
          <span>
            Friday, 18 October 2019 at 11:45
          </span>
        </footer>
      </article>
    </div>
  </div>
</div>
`;
